The JDK 6 and 7 builds are broken

An example log can be found here:
https://scala-webapps.epfl.ch/jenkins/job/scalajs-task-worker/213646/consoleFull
I have also archived it in the following gist, just to make sure it doesn't disappear:
https://gist.github.com/sjrd/53d5c65f8da54f97c8b98aa5fdffd401

Fetching the Scala sources from the GitHub git through HTTPS has started failing two weeks ago, it seems. The failure is:
```
[info] Fetching Scala source version 2.11.12
org.eclipse.jgit.api.errors.TransportException: https://github.com/scala/scala.git: cannot open git-upload-pack
	at org.eclipse.jgit.api.FetchCommand.call(FetchCommand.java:137)
	at org.eclipse.jgit.api.CloneCommand.fetch(CloneCommand.java:178)
	at org.eclipse.jgit.api.CloneCommand.call(CloneCommand.java:125)
	at build.Build$$anonfun$partest$1$$anonfun$apply$183.apply(Build.scala:1750)
	at build.Build$$anonfun$partest$1$$anonfun$apply$183.apply(Build.scala:1733)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:334)
	at java.util.concurrent.FutureTask.run(FutureTask.java:166)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:334)
	at java.util.concurrent.FutureTask.run(FutureTask.java:166)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1110)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:603)
	at java.lang.Thread.run(Thread.java:722)
Caused by: org.eclipse.jgit.errors.TransportException: https://github.com/scala/scala.git: cannot open git-upload-pack
	at org.eclipse.jgit.transport.TransportHttp.connect(TransportHttp.java:499)
	at org.eclipse.jgit.transport.TransportHttp.openFetch(TransportHttp.java:308)
	at org.eclipse.jgit.transport.FetchProcess.executeImp(FetchProcess.java:136)
	at org.eclipse.jgit.transport.FetchProcess.execute(FetchProcess.java:122)
	at org.eclipse.jgit.transport.Transport.fetch(Transport.java:1111)
	at org.eclipse.jgit.api.FetchCommand.call(FetchCommand.java:128)
	at org.eclipse.jgit.api.CloneCommand.fetch(CloneCommand.java:178)
	at org.eclipse.jgit.api.CloneCommand.call(CloneCommand.java:125)
	at build.Build$$anonfun$partest$1$$anonfun$apply$183.apply(Build.scala:1750)
	at build.Build$$anonfun$partest$1$$anonfun$apply$183.apply(Build.scala:1733)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:334)
	at java.util.concurrent.FutureTask.run(FutureTask.java:166)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:334)
	at java.util.concurrent.FutureTask.run(FutureTask.java:166)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1110)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:603)
	at java.lang.Thread.run(Thread.java:722)
Caused by: javax.net.ssl.SSLException: Received fatal alert: protocol_version
	at sun.security.ssl.Alerts.getSSLException(Alerts.java:208)
	at sun.security.ssl.Alerts.getSSLException(Alerts.java:154)
	at sun.security.ssl.SSLSocketImpl.recvAlert(SSLSocketImpl.java:1943)
	at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:1059)
	at sun.security.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1294)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1321)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1305)
	at sun.net.www.protocol.https.HttpsClient.afterConnect(HttpsClient.java:515)
	at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:185)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1299)
	at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:468)
	at sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:338)
	at org.eclipse.jgit.util.HttpSupport.response(HttpSupport.java:167)
	at org.eclipse.jgit.transport.TransportHttp.connect(TransportHttp.java:462)
	at org.eclipse.jgit.transport.TransportHttp.openFetch(TransportHttp.java:308)
	at org.eclipse.jgit.transport.FetchProcess.executeImp(FetchProcess.java:136)
	at org.eclipse.jgit.transport.FetchProcess.execute(FetchProcess.java:122)
	at org.eclipse.jgit.transport.Transport.fetch(Transport.java:1111)
	at org.eclipse.jgit.api.FetchCommand.call(FetchCommand.java:128)
	at org.eclipse.jgit.api.CloneCommand.fetch(CloneCommand.java:178)
	at org.eclipse.jgit.api.CloneCommand.call(CloneCommand.java:125)
	at build.Build$$anonfun$partest$1$$anonfun$apply$183.apply(Build.scala:1750)
	at build.Build$$anonfun$partest$1$$anonfun$apply$183.apply(Build.scala:1733)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:334)
	at java.util.concurrent.FutureTask.run(FutureTask.java:166)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:334)
	at java.util.concurrent.FutureTask.run(FutureTask.java:166)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1110)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:603)
	at java.lang.Thread.run(Thread.java:722)
[error] (partest/*:fetchScalaSource) org.eclipse.jgit.api.errors.TransportException: https://github.com/scala/scala.git: cannot open git-upload-pack
[error] Total time: 1 s, completed Mar 2, 2018 2:34:28 AM
```
I have traced this issue to https://stackoverflow.com/q/34283829/1829647 and the Oracle blog post at https://blogs.oracle.com/java-platform-group/diagnosing-tls,-ssl,-and-https. The bottom line is that JDK 6 and 7 do not support TLS v1.2.

I suspect that GitHub has changed their server configuration so that its HTTPS git server refuses to operate in TLS v1.0 or v1.1. This means that JGit has no hope to communicate with GitHub repos on JDK 6 and 7 anymore.

The proper fix would be to shell out and use the command-line `git` client to fetch the Scala sources. But right now we need to disable the JDK 6 and 7 builds to restore the CI.

--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -426,7 +426,7 @@ def Tasks = [
 ]
 
 def mainJavaVersion = "1.8"
-def otherJavaVersions = [] // should be ["1.6", "1.7"] but that's broken (see #3293)
+def otherJavaVersions = []
 def allJavaVersions = otherJavaVersions.clone()
 allJavaVersions << mainJavaVersion
 
@@ -473,9 +473,7 @@ mainScalaVersions.each { scalaVersion ->
   }
   quickMatrix.add([task: "bootstrap", scala: scalaVersion, java: mainJavaVersion])
   if (!scalaVersion.startsWith("2.10.")) {
-    // #3293, should be: def javaVersion = scalaVersion.startsWith("2.11.") ? "1.7" : mainJavaVersion
-    def javaVersion = "1.8"
-    quickMatrix.add([task: "partest-fastopt", scala: scalaVersion, java: javaVersion])
+    quickMatrix.add([task: "partest-fastopt", scala: scalaVersion, java: mainJavaVersion])
   }
 }
 noToolsScalaVersions.each { scalaVersion ->
@@ -485,14 +483,10 @@ noToolsScalaVersions.each { scalaVersion ->
 }
 allJavaVersions.each { javaVersion ->
   quickMatrix.add([task: "tools-cli-stubs-sbtplugin", scala: "2.10.7", java: javaVersion])
-  if (javaVersion != "1.6") {
-    // Tools do not compile on JDK6, Scala 2.11.x (see #1235)
-    quickMatrix.add([task: "tools-cli-stubs", scala: "2.11.12", java: javaVersion])
-  }
 }
 quickMatrix.add([task: "tools-cli-stubs", scala: "2.12.8", java: mainJavaVersion])
-quickMatrix.add([task: "partestc", scala: "2.11.0", java: "1.8"]) // #3293, should be: 1.7
-quickMatrix.add([task: "sbtplugin-test", toolsscala: "2.10.7", sbt_version_override: "", java: "1.7"])
+quickMatrix.add([task: "partestc", scala: "2.11.0", java: mainJavaVersion])
+quickMatrix.add([task: "sbtplugin-test", toolsscala: "2.10.7", sbt_version_override: "", java: mainJavaVersion])
 quickMatrix.add([task: "sbtplugin-test", toolsscala: "2.12.8", sbt_version_override: "1.0.0", java: mainJavaVersion])
 
 // The 'full' matrix

--- a/project/Build.scala
+++ b/project/Build.scala
@@ -353,11 +353,8 @@ object Build {
       scalacOptions in (Compile, doc) := {
         val baseOptions = (scalacOptions in (Compile, doc)).value
 
-        /* - need JDK7 to link the doc to java.nio.charset.StandardCharsets
-         * - in Scala 2.10, some ScalaDoc links fail
-         */
-        val fatalInDoc =
-          javaVersion.value >= 7 && scalaBinaryVersion.value != "2.10"
+        // in Scala 2.10, some ScalaDoc links fail
+        val fatalInDoc = scalaBinaryVersion.value != "2.10"
 
         if (fatalInDoc) baseOptions
         else baseOptions.filterNot(_ == "-Xfatal-warnings")
@@ -469,6 +466,8 @@ object Build {
         val fullVersion = System.getProperty("java.version")
         val v = fullVersion.stripPrefix("1.").takeWhile(_.isDigit).toInt
         sLog.value.info(s"Detected JDK version $v")
+        if (v < 8)
+          throw new MessageOnlyException("This build requires JDK 8 or later. Aborting.")
         v
       }
   )
@@ -1544,9 +1543,8 @@ object Build {
       val isScalaAtLeast212 =
         !scalaV.startsWith("2.10.") && !scalaV.startsWith("2.11.")
 
-      List(sharedTestDir / "scala") ++
-      includeIf(sharedTestDir / "require-jdk7", javaVersion.value >= 7) ++
-      includeIf(sharedTestDir / "require-jdk8", javaVersion.value >= 8) ++
+      List(sharedTestDir / "scala", sharedTestDir / "require-jdk7",
+          sharedTestDir / "require-jdk8") ++
       includeIf(testDir / "require-2.12", isJSTest && isScalaAtLeast212)
     },
 

